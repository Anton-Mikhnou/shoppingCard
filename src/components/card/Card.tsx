import ButtonAdd from '../../UI/buttonAdd/ButtonAdd'
import ButtonBuy from '../../UI/buttonBuy/ButtonBuy'
import ButtonDiminish from '../../UI/buttonDiminish/buttonDiminish'
import style from './Card.module.scss'
import { useAppDispatch } from '../../app/hooks'
import { increment } from '../../modules/counterSlice' 
import { useState } from 'react'
// import { selectCart } from '../modules/cartSlice'
import { addToCart } from '../../modules/cartSlice'

interface CardProps {
    name: string;
    url: string;
    price: number;
    id: number;
}

export default function Card({name, url, price, id }: CardProps) {
    const [quality, setQuality] = useState<number>(0);
    const dispatch = useAppDispatch();

    const handelInputChange = (event: React.FormEvent<HTMLInputElement>): void => {
        const value: number = parseInt(event.currentTarget.value, 10) || quality;

        setQuality(value);
    }

    function handleSabtractCount() {
        if(quality === 0) {
            return
        }
        setQuality(quality => quality - 1);        
    }

    function handleAddCount() {
        setQuality(quality => quality + 1);        
    }
    
    function handleBuy() {
        dispatch(increment(quality));
        dispatch(addToCart({id: id, quality:quality}))
    }
    
    return (
        <>
            <div className={style['wrapper']}>
                <div className={style['container']}>
                    <div className={style['container__img']}>
                        <img src={url} alt="" />
                    </div>
                    <div className={style['description']}>
                        <h2>{name}</h2>
                        <h3>${price}</h3>
                        <div className={style['description__buttons']}>
                            <div>
                                <ButtonDiminish handleSabtractCount={handleSabtractCount}/>
                                <input type="number"  value={quality} onChange={handelInputChange}/>
                                <ButtonAdd handleAddCount={handleAddCount}/>
                            </div>
                            <ButtonBuy handleBuy={handleBuy}/>
                        </div>
                    </div>
                </div>
            </div>
        </>
    )
}