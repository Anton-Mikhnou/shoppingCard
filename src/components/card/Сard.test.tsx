import { render, screen} from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import cartReducer from "../../modules/cartSlice"; 
import Card from "./Card";

const mockStore = configureStore({
    reducer: {
      cart: cartReducer,
    },
});


describe('Card component', () => {
    it('render card', () => {
    
        render(
            <Provider store={mockStore}>
                <Card title="Product" price={100} url="../../public/img/970743.webp" id={1}/>
            </Provider>
        );

        const headingTwo = screen.getByRole('heading',{level: 2});
        expect(headingTwo).toHaveTextContent(/product/i);

        const headintTrird = screen.getByRole('heading', {level: 3});
        expect(headintTrird).toHaveTextContent(/100/i);

        const img = screen.getByRole('img');
        expect(img).toHaveAttribute('src', '../../public/img/970743.webp');


        screen.debug();
        
    })
})

describe('Card component function', () => {
    it('')
})
