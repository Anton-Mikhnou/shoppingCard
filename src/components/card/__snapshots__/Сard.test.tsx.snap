// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > render card 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_wrapper_c39673"
      >
        <div
          class="_container_c39673"
        >
          <div
            class="_container__img_c39673"
          >
            <img
              alt=""
              src="../../public/img/970743.webp"
            />
          </div>
          <div
            class="_description_c39673"
          >
            <h2>
              hi
            </h2>
            <h3>
              $
              100
            </h3>
            <div
              class="_description__buttons_c39673"
            >
              <div>
                <button
                  class="_buttonDiminish_174586"
                >
                  -
                </button>
                <input
                  type="number"
                  value="0"
                />
                <button
                  class="_buttonAdd_54f96f"
                >
                  +
                </button>
              </div>
              <button
                class="_addToCart_a9c296"
              >
                Add to Cart
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_wrapper_c39673"
    >
      <div
        class="_container_c39673"
      >
        <div
          class="_container__img_c39673"
        >
          <img
            alt=""
            src="../../public/img/970743.webp"
          />
        </div>
        <div
          class="_description_c39673"
        >
          <h2>
            hi
          </h2>
          <h3>
            $
            100
          </h3>
          <div
            class="_description__buttons_c39673"
          >
            <div>
              <button
                class="_buttonDiminish_174586"
              >
                -
              </button>
              <input
                type="number"
                value="0"
              />
              <button
                class="_buttonAdd_54f96f"
              >
                +
              </button>
            </div>
            <button
              class="_addToCart_a9c296"
            >
              Add to Cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
